crates/aiken-project/src/blueprint/validator.rs

```diff
@@ -116,62 +118,64 @@ impl Validator {
+                schema,
+            });
+
+        let redeemer = Annotated::from_type(
+            modules.into(),
+            tipo_or_annotation(module, redeemer),
+            &mut definitions,
+        )
+        .map_err(|error| Error::Schema {
+            error,
+            location: redeemer.location,
+            source_code: NamedSource::new(
+                module.input_path.display().to_string(),
+                module.code.clone(),
+            ),
+        })
+        .map(|schema| Parameter {
+            title: Some(redeemer.arg_name.get_label()),
+            schema: match datum {
+                Some(..) if is_multi_validator => {
+                    Annotated::as_wrapped_redeemer(&mut definitions, schema, redeemer.tipo.clone())
+                }
+                _ => schema,
+            },
+        })?;
+
         Ok(Validator {
             title: format!("{}.{}", &module.name, &func.name),
             description: func.doc.clone(),
             parameters,
-            datum: datum
-                .map(|datum| {
-                    Annotated::from_type(
-                        modules.into(),
-                        tipo_or_annotation(module, datum),
-                        &mut definitions,
-                    )
-                    .map_err(|error| Error::Schema {
-                        error,
-                        location: datum.location,
-                        source_code: NamedSource::new(
-                            module.input_path.display().to_string(),
-                            module.code.clone(),
-                        ),
-                    })
-                })
-                .transpose()?
-                .map(|schema| Parameter {
-                    title: datum.map(|datum| datum.arg_name.get_label()),
-                    schema,
-                }),
-            redeemer: Annotated::from_type(
-                modules.into(),
-                tipo_or_annotation(module, redeemer),
-                &mut definitions,
-            )
-            .map_err(|error| Error::Schema {
-                error,
-                location: redeemer.location,
-                source_code: NamedSource::new(
-                    module.input_path.display().to_string(),
-                    module.code.clone(),
-                ),
-            })
-            .map(|schema| Parameter {
-                title: Some(redeemer.arg_name.get_label()),
-                schema: match datum {
-                    Some(..) if is_multi_validator => Annotated::as_wrapped_redeemer(
-                        &mut definitions,
-                        schema,
-                        redeemer.tipo.clone(),
-                    ),
-                    _ => schema,
-                },
-            })?,
-            program: program.clone(),
+            datum,
+            redeemer,
+            program: program.get(generator, def, &module.name),
```

Comment:
Not sure to get the rationale for the memoization thing? I have a sense that it isn't only for performance but also solving something else?

Reply:
I literally just didn't want to call generate twice, it felt like an extreme waste of computation. Especially for largely project with many validators. We essentially double the amount of computation regarded.

Reply:
But where was it called _twice_ 😅 (genuine ask)?

Reply:
It wasn't before but that was because it was done before validations at the top of `from_checked_modules`, therefore causing a panic before some validations could return a proper error. Code gen will panic if the validator has a free var for example. I then had to delay generation and so that means it must happen in the secondary utility method. Which can potentially be called twice in case of a multi-validator. Which meant generation was potentially happening twice. With this little setup we call generate exactly once every time regardless.